pinot-common/src/main/java/org/apache/pinot/common/datablock/BaseDataBlock.java

```diff
@@ -426,168 +401,202 @@ public Map<Integer, String> getExceptions() {
     return _errCodeToExceptionMap;
   }
 
-  /**
-   * Serialize this data block to a byte array.
-   * <p>
-   * In order to deserialize it, {@link DataBlockUtils#getDataBlock(ByteBuffer)} should be used.
-   */
   @Override
-  public byte[] toBytes()
+  public List<ByteBuffer> serialize()
       throws IOException {
-    ThreadResourceUsageProvider threadResourceUsageProvider = new ThreadResourceUsageProvider();
-
-    UnsynchronizedByteArrayOutputStream byteArrayOutputStream = new UnsynchronizedByteArrayOutputStream(8192);
-    DataOutputStream dataOutputStream = new DataOutputStream(byteArrayOutputStream);
-    writeLeadingSections(dataOutputStream);
-
-    // Write metadata: length followed by actual metadata bytes.
-    // NOTE: We ignore metadata serialization time in "responseSerializationCpuTimeNs" as it's negligible while
-    // considering it will bring a lot code complexity.
-    serializeMetadata(dataOutputStream);
-
-    return byteArrayOutputStream.toByteArray();
-  }
-
-  private void writeLeadingSections(DataOutputStream dataOutputStream)
-      throws IOException {
-    dataOutputStream.writeInt(getDataBlockVersionType());
-    dataOutputStream.writeInt(_numRows);
-    dataOutputStream.writeInt(_numColumns);
-    int dataOffset = HEADER_SIZE;
-
-    // Write exceptions section offset(START|SIZE).
-    dataOutputStream.writeInt(dataOffset);
-    byte[] exceptionsBytes;
-    exceptionsBytes = serializeExceptions();
-    dataOutputStream.writeInt(exceptionsBytes.length);
-    dataOffset += exceptionsBytes.length;
-
-    // Write dictionary map section offset(START|SIZE).
-    dataOutputStream.writeInt(dataOffset);
-    byte[] dictionaryBytes = null;
-    if (_stringDictionary != null) {
-      dictionaryBytes = serializeStringDictionary();
-      dataOutputStream.writeInt(dictionaryBytes.length);
-      dataOffset += dictionaryBytes.length;
-    } else {
-      dataOutputStream.writeInt(0);
-    }
-
-    // Write data schema section offset(START|SIZE).
-    dataOutputStream.writeInt(dataOffset);
-    byte[] dataSchemaBytes = null;
-    if (_dataSchema != null) {
-      dataSchemaBytes = _dataSchema.toBytes();
-      dataOutputStream.writeInt(dataSchemaBytes.length);
-      dataOffset += dataSchemaBytes.length;
-    } else {
-      dataOutputStream.writeInt(0);
+    if (_serialized == null) {
+      _serialized = DataBlockUtils.serialize(this);
```

Comment:
I think the opposite. Having serdes as separate classes:
1. Simplifies the class. BaseDataBlock is very complex, by removing the serialization/deserialization code we simplify this class.
2. Separates responsibilities. 
   * DataBlock defines how to use the block in the OpChain. On a cluster running different versions of Pinot, the memory layout of a data block may be different. 
   * DataBlockSerde defines the layout used to send the blocks through the network. Different Pinot instances may support different block versions, but in order to be able to talk to each other they have to have at least one version in common.
3. Opens the possibility of having different implementations.
   * Imagine BaseDataBlock includes the serde code. If we want to try a new implementation, we need to modify the code, destroying the older code. That means it is more difficult to test or benchmark the new version.
   * In the first commits of this PR I included a class that implemented the V1_V2 protocol version but using almost the same code we use right now in master. I also kept the old `toBytes` code. It was easy to benchmark each implementation by having these options. I decided to remove these classes to make this PR simpler, but during development it was useful.
   * Imagine we create a new serde for the current V1_V2 version that is more efficient. The code does not support it right now but it is very easy to be able to configure which serde we want to use for each version with PinotConfiguration. This means we can try to use one serde by default but change to the other in case we find a bug.

> I think keeping it within the class is easier for version control.

I don't understand this. Obviously this PR destroys git history. But once we start using this mechanism there will be no difference in terms of git.

Imagine that we add a V3 that serializes blocks using Apache Arrow or Parquet. V1_V2 and this theoretical V3 would be super different.

Reply:
By version control, I meant the version of `DataBlock`.
Seems like the idea in this PR is to decouple the `DataBlock` version and `DataBlockSerDe` version, which is counter intuitive to me. Even if we want to maintain multiple ser-de versions so that we can try out new implementations, they should be under the same `DataBlock` version, where the serialized bytes remains the same. Say if we introduce a new `DataBlock` version with different bytes representation on the wire, I don't see how we can reuse the same `DataBlockSerDe` for it.

Reply:
> Seems like the idea in this PR is to decouple the DataBlock version and DataBlockSerDe version, which is counter intuitive to me.

`DataBlock` has been used as the way data is represented in memory during `OpChain` execution AND the layout used to send data from one server to the other. That is what I'm trying to break here. My proposal is that 
`DataBlock` should be the object we use represent data in memory in a `OpChain`. The way we serialize that memory layout is independent and it is defined by the `DataBlockSerDe` version.

Two servers in two different Pinot versions can have a different memory representation of the same information. The only serde we have right now mostly uses the same bytes in memory, so the memory and network representation is almost the same. But that may not be the case in the future.

Another usage is in case we want to test a new serialization implementation that is binary compatible with the older one. We can have two `DataBlockSerDe` implementations for the same version and we can swap between one or the other whenever we want. This opens the ability to change the implementation used with a Pinot configuration but probably more useful, makes it easier to create a benchmark and to create tests that verify both implementations are compatible (we can create a unit test that serializes with one implementation and deserializes with the other).

> Even if we want to maintain multiple ser-de versions so that we can try out new implementations, they should be under the same DataBlock version, where the serialized bytes remains the same.

My suggestion is to forget about DataBlock version. If in future we have newer ways to represent a DataBlock (ie using Apache Arrow) we will need to define how to serialize and deserialize that new block, but we won't need to modify the data block version. In fact `DataBlock.getVersion()` and `BaseDataBlock.getDataBlockVersionType()` have been removed.

> Say if we introduce a new DataBlock version with different bytes representation on the wire, I don't see how we can reuse the same DataBlockSerDe for it.

That is not possible in this definition because the `DataBlock` do not define the wire representation.

We could:
1. Create a new `DataBlock` that is serialized as one of the older ones in protocol version `V1_V2`. This will be probably be expensive, but it is required to support older versions.
2. Create a new `DataBlock` that is serialized and deserialized by a new `SerDe` in version `V3`. This will be more efficient, but can only be used to communicate two servers that understand this version.
3. Create a new `SerDe` that generates different wire representation for the `DataBlocks` we already have. This new `SerDe` should use a new version.
4. Create a new `SerDe` that generates the same wire representation for the `DataBlocks` we already have. This new `SerDe` could use the current `V1_V2` version.

Reply:
Thanks for the explanation. Now I understand why the version is moved into the ser-de class: basically we want to shift the wire format control into the `DataBlockSerDe`, and `DataBlock` only defines the in-memory format right?
Since you are removing the version from `DataBlock`, and currently we only allow a single version of `DataBlock`, should we consider also cleaning up the backward compatible `DataBlock`s in the test since this no longer apply? In the `DataBlockSerDe`, we can make a clean start from version 0 or 1 to reduce confusion for future developers.

Reply:
That makes sense. I don't know if I'm going to have time this week to apply the change, but I'll try to do that as soon as possible.